[TASK-11355] fix: token amount input

jjramirezn opened the pull request:
Refactor token amount input to use a more robust approach.

The component communicates two values: the token amount and the USD value. Also keeps two display values internally that it shows: the main display amount and the alternative amount, which depending on the mode that the input is using will be between token and usd or fiat and usd

jjramirezn committed:
fix: token amount input

Refactor token amount input to use a more robust approach.

The component communicates two values: the token amount and the USD value. Also keeps two display values internally that it shows: the main display amount and the alternative amount, which depending on the mode that the input is using will be between token and usd or fiat and usd

jjramirezn committed:
refactor: safely parse walletbalance

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

jjramirezn committed:
style: Apply prettier formatting

jjramirezn commented on src/components/Payment/PaymentForm/index.tsx:
```diff
@@ -252,9 +278,19 @@
+        ) {
+            tokenAmount = (parseFloat(inputUsdValue) / requestedTokenPrice).toString()
+        }
+
```
good catch

jjramirezn committed:
refactor: avoid division by zero

jjramirezn commented on src/components/Global/TokenAmountInput/index.tsx:
```diff
@@ -27,64 +26,141 @@ const TokenAmountInput = ({
     const inputRef = useRef<HTMLInputElement>(null)
     const inputType = useMemo(() => (window.innerWidth < 640 ? 'text' : 'number'), [])
-    const { isConnected: isPeanutWallet } = useWallet()
-
-    // If currency is specified, force CUSTOM mode and prevent switching
-    const [currencyMode, setCurrencyMode] = useState<'USD' | 'CUSTOM'>(currency ? 'CUSTOM' : 'USD')
 
```
nah

jjramirezn commented on src/components/Global/TokenAmountInput/index.tsx:
```diff
@@ -27,64 +26,141 @@
+            setTokenValue(tokenValue)
+        },
+        [displayMode, currency?.price, selectedTokenData?.price, calculateAlternativeValue]
+    )
```
thankyou

jjramirezn committed:
refactor: switch case to use braces

jjramirezn commented on src/components/Global/TokenAmountInput/index.tsx:
```diff
@@ -148,89 +212,29 @@
-                                // Going from custom currency to USD
-                                setCurrencyMode('USD')
-                                // Display the actual token value (USD)
-                                setDisplayValue(tokenValue || '')
-                            }
-                        } else if (selectedTokenData?.price) {
-                            setInputDenomination(inputDenomination === 'USD' ? 'TOKEN' : 'USD')
-                        }
+                        const currentValue = displayValue
+                        setDisplayValue(alternativeDisplayValue)
+                        setAlternativeDisplayValue(currentValue)
+                        setIsInputUsd(!isInputUsd)
                     }}
                 >
```
Toggling does not change the underlaying tokenValue, it's the same so it makes sense that it doesnt change when toggling. The only thing that changes is the displayValue to be shown

jjramirezn commented on src/components/Payment/PaymentForm/index.tsx:
```diff
@@ -132,24 +138,49 @@ export const PaymentForm = ({
-                    setUsdValue(usdValue)
+                    setRequestedTokenPrice(priceData.price)
+
+                    if (requestDetails?.tokenAmount) {
+                        // calculate USD value
+                        const tokenAmount = parseFloat(requestDetails.tokenAmount)
+                        const usdValue = formatAmount(tokenAmount * priceData.price)
+                        setInputTokenAmount(usdValue)
+                        setUsdValue(usdValue)
+                    }
```
nope

jjramirezn commented on src/components/Payment/PaymentForm/index.tsx:
```diff
@@ -305,15 +349,10 @@ export const PaymentForm = ({
-            }
-        } else if (inputDenomination === 'USD') {
-            setUsdValue(inputTokenAmount)
+        if (selectedTokenPrice) {
+            setUsdValue((parseFloat(inputTokenAmount) * selectedTokenPrice).toString())
         }
-    }, [inputTokenAmount, inputDenomination, selectedTokenPrice])
+    }, [inputTokenAmount, selectedTokenPrice])
```
aha

jjramirezn commented on src/components/Global/TokenAmountInput/index.tsx:
```diff
@@ -27,64 +26,149 @@ const TokenAmountInput = ({
+            }
+        },
+        [displayMode, currency?.price, selectedTokenData?.price, isInputUsd, decimals]
+    )
```
o.o

Hugo0 commented on src/components/Cashout/Components/Initial.view.tsx:
```diff
@@ -204,24 +196,10 @@ export const InitialCashoutView = ({
         }
     }
 
-    const maxValue = useMemo(() => {
-        return printableUsdc(balance)
-    }, [balance])
-
```
nicely removed

Hugo0 commented on src/components/Global/TokenAmountInput/index2_broken.tsx:
thx :)

Hugo0 commented on src/components/Request/link/views/Create.request.link.view.tsx:
```diff
@@ -62,9 +61,7 @@ export const CreateRequestLinkView = () => {
-    const [_tokenValue, _setTokenValue] = useState<string>(
-        (inputDenomination === 'TOKEN' ? tokenValue : usdValue) ?? ''
-    )
+    const [_tokenValue, _setTokenValue] = useState<string>(tokenValue ?? '')
```
:+1:

Hugo0 commented on src/components/Request/link/views/Create.request.link.view.tsx:
```diff
@@ -166,18 +160,11 @@ export const CreateRequestLinkView = () => {
-            if (selectedTokenPrice) {
-                setTokenValue((parseFloat(_tokenValue) / selectedTokenPrice).toString())
-            }
+        setTokenValue(_tokenValue)
+        if (selectedTokenPrice) {
+            setUsdValue((parseFloat(_tokenValue) * selectedTokenPrice).toString())
```
this is a much cleaner approach :+1:

Hugo0 commented on src/context/tokenSelector.context.tsx:
```diff
@@ -15,13 +15,11 @@ import {
 import { useWallet } from '@/hooks/wallet/useWallet'
 import { type ITokenPriceData } from '@/interfaces'
-import { estimateIfIsStableCoinFromPrice, getUserPreferences } from '@/utils'
+import { getUserPreferences } from '@/utils'
```
thought: we might want to store usd vs token denomination in user preferences too (some users might consistnetly prefer one over the other)

Hugo0 commented on src/components/Payment/PaymentForm/index.tsx:
```diff
@@ -132,24 +138,49 @@ export const PaymentForm = ({
+        }
+
+        if (isActivePeanutWallet && areEvmAddressesEqual(selectedTokenAddress, PEANUT_WALLET_TOKEN)) {
+            const walletNumeric = parseFloat(String(peanutWalletBalance).replace(/,/g, ''))
+            const inputNumeric = parseFloat(String(inputTokenAmount).replace(/,/g, ''))
```
fyi this will lose precision for 18+ digits. is that a problem for request fulfillment? what happens again if a request is for 1.0 ETH and we pay 0.9999999999999999999999?

Hugo0 replied:
iirc that's not valid and we needed +=

Hugo0 replied:
(however for payment fulfillment we dont ahve to worry about input, we set that ourselves)

jjramirezn replied:
I think this is ok for a first check (this is to check balance) and in those cases It will fail on attempt

Hugo0 commented on src/components/Global/TokenAmountInput/index.tsx:
```diff
@@ -27,64 +26,149 @@ const TokenAmountInput = ({
+            _decimals = PEANUT_WALLET_TOKEN_DECIMALS
         }
-    }
+        // For displaying the token amount, anything more breaks the UI
```
0.000001 btc = 10cts

i think thats fine

jjramirezn replied:
Or we denominate in sats 1000 sats = 1 usd and btc apps are using it

Hugo0 commented on src/components/Global/TokenAmountInput/index.tsx:
```diff
@@ -27,64 +26,149 @@ const TokenAmountInput = ({
+
+    const calculateAlternativeValue = useCallback(
+        (value: string) => {
+            // There is no alternative display when dealing with stables
```
v nice comment!

Hugo0 commented on src/components/Global/TokenAmountInput/index.tsx:
```diff
@@ -27,64 +26,149 @@ const TokenAmountInput = ({
+        switch (displayMode) {
+            case 'STABLE': {
+                setDisplaySymbol('$')
+                setAlternativeDisplaySymbol('')
```
lovely to have alternative display :+1:

Hugo0 reviewed (approved):
see discord comments

nice!

jjramirezn committed:
fix: token amount input

- Enable currency mode for non stablecoin tokens
- Fix input decimals and accepted characters
